### PR TITLE
Fix Matrix Equality and MatrixExpr

### LINKED ISSUE #7842: `Equality` and `MatrixExpr`
There are a couple issues here, just not sure about the best way to fix them.

A couple things to setup the examples:

```
>>> A = MatrixSymbol('A', 3, 1)
>>> B = MatrixSymbol('B', 2, 1)
```

1.) A and B cannot be equal, but Symbol doesn't care:

```
>>> Eq(A, B)
A == B
```

I think this could be fixed by giving `MatrixSymbol` its own `_eval_Eq` method, as `ImmutableMatrix` does, but this method could also go in `MatrixExpr` itself. The second is probably preferrable, as it would work for anything containing `MatrixExpr`

2.) Individual elements should be able to be set equivalent (as you can with Symbol), but you can't:

```
>>> Eq(A[1,0], B[1, 0])
# Trimming long error
/home/jimmy/Code/sympy/sympy/core/expr.py in equals(self, other, failing_expression)
    554         # because if the expression ever goes to 0 then the subsequent
    555         # simplification steps that are done will be very fast.
--> 556         diff = factor_terms((self - other).simplify(), radical=True)
    557 
    558         if not diff:

/home/jimmy/Code/sympy/sympy/core/expr.py in simplify(self, ratio, measure)
   2882         from sympy.core.function import count_ops
   2883         measure = measure or count_ops
-> 2884         return simplify(self, ratio, measure)
   2885 
   2886     def nsimplify(self, constants=[], tolerance=None, full=False):

/home/jimmy/Code/sympy/sympy/simplify/simplify.py in simplify(expr, ratio, measure, fu)
   3692     _e = cancel(expr)
   3693     expr1 = shorter(_e, _mexpand(_e).cancel())  # issue 6829
-> 3694     expr2 = shorter(together(expr, deep=True), together(expr1, deep=True))
   3695 
   3696     if ratio is S.Infinity:

/home/jimmy/Code/sympy/sympy/simplify/simplify.py in shorter(*choices)
   3686         if not has_variety(choices):
   3687             return choices[0]
-> 3688         return min(choices, key=measure)
   3689 
   3690     expr = bottom_up(expr, lambda w: w.normal())

/home/jimmy/Code/sympy/sympy/core/function.py in count_ops(expr, visual)
   2263         while args:
   2264             a = args.pop()
-> 2265             if a.is_Rational:
   2266                 #-1/3 = NEG + DIV
   2267                 if a is not S.One:

AttributeError: 'str' object has no attribute 'is_Rational'
```

I have two thoughts on fixing this:
1. Make `MatrixElement` a subclass of Symbol. Perhaps this is a bad idea, but then it would work with the current `Eq` method.
2. Add `MatrixElement` to the `has` statement in `_eval_sides` method of `Relational` in sympy/core/Relational.py. This seems like a hackish thing though.

I think making MatrixElement a subclass of Symbol is the best bet, but doing that change results in tons of errors right now. Thoughts?

### PR DESCRIPTION
**Pull request by divyanshu132**
#### References to other Issues or PRs
Fixes #7842 
Fixes #16042

#### Brief description of what is fixed or changed
Test case for equality in matrices.

#### Release Notes

*  matrices
   *  _eval_Eq added for MatExpr

### CODE REVIEW

**Comment by divyanshu132** [comment]
@oscarbenjamin I've added the additional tests that you pointed here https://github.com/sympy/sympy/pull/16042#discussion_r259276424 please review.

File: sympy/matrices/expressions/tests/test_matexpr.py [hunk 1]
```diff
@@ -360,3 +360,14 @@ def test_issue_2749():
 def test_issue_2750():
     x = MatrixSymbol('x', 1, 1)
     assert (x.T*x).as_explicit()**-1 == Matrix([[x[0, 0]**(-2)]])
+
+
+def test_matrix_equality():
+    from sympy import Eq
```
Comment (smichr) [inline sympy/matrices/expressions/tests/test_matexpr.py]: import Eq at top and please refer to issue number: as I noted on the closed PR, if this ever fails it gives an easy reference to the original issue whereas the subject of the test -- matrix equality -- is self evident from the tests added.

**Comment by oscarbenjamin** [comment]
Can you push these changes to #16042 instead of opening a new PR?

Opening a new PR loses all the discussion that has happened in the previous one.

**Comment by divyanshu132** [comment]
@oscarbenjamin actually the previous pr had some merge conflicts that's why I've deleted that branch.

File: sympy/matrices/expressions/tests/test_matexpr.py (hunk 1 above)
Comment (divyanshu132) [reply sympy/matrices/expressions/tests/test_matexpr.py]: okay I'll do that.

**Pull request merged**